components/sync15/src/engine/bridged_engine.rs

```diff
@@ -2,8 +2,9 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
-use crate::bso::{IncomingBso, OutgoingBso};
+use anyhow::Result;
```

Comment:
I think that could work! The tricky part before was the `sync15` / `sync15_traits` split, so we didn't have a common error type to use—I think that's why we used the associated `Error` type. As a bonus, the associated `Error` type also made it "easier" to turn a bridged engine error into something we could bubble up through Golden Gate, before turning it into an `nsresult`.

But now that @mhammond has merged `sync15` and `sync15_traits`, and the "bonus" of converting to `nsresult` turned out to not be helpful, I think it should be easier to use `sync15::Error` down the road!

Reply:
I'm not sure I fully understand how this would be useful. Consider, eg, places:
* The places sync code works almost exclusively with `places::Error` (ie, errors returned by the impl will always be `places::Error` and converted to anyhow right at the trait boundary)
* Places already has an [Into for `sync15::Error`](https://github.com/mozilla/application-services/blob/main/components/places/src/error.rs#L59)

Because `sync15` doesn't depend on places (and indeed, doesn't know how many engines might exist), I think the only sane way of doing this would be to have a new sync15::Error variant with, say `EngineError(String)` which all engines would need to use to convert their specific errors into a `sync15:Error` - which is just `anyhow:Error` but with more steps :)

Further, this would mean you can go from a `sync15::Error` into a `places::Error` *and* a way to go from a `places::Error` into a `sync15::Error` - which seems a little odd and would be impossible to round-trip sanely - ie, I don't think you could go from `places::Error` -> `sync15::Error` -> `places::Error`.

Although there's a good chance I'm misunderstanding something here :)

Reply:
Ooof I didn't realize the engines already had `From<sync15::Error>` implementations :(

> , I think the only sane way of doing this would be to have a new sync15::Error variant with, say EngineError(String) which all engines would need to use to convert their specific errors into a sync15:Error - which is just anyhow:Error but with more steps :)

Yeah that doesn't sound great :( It might be possible to define **YET ANOTHER** error type in `sync15` that is meant to be a target for the SyncEngine traits.. but it's starting to get messy...

 I concede, I don't see anything wrong with `anyhow` **unless consumers want to handle specific errors without deref'ing**, which I don't think any consumers do


As a slightly related question: Why does `places::Error` implement `From<sync15::Error>`, I imagine it's because of the "stand-alone" sync we support for iOS's sake?

Reply:
> I imagine it's because of the "stand-alone" sync we support for iOS's sake?

Yep, that's exactly why, so I guess eventually what I wrote above need not be true.

But as you imply above, nothing in `sync15` can possibly handle a specific `places` error because (a) it doesn't depend places and (b) given it doesn't even know which engines might exist it doesn't really make sense. If places ever did need to see a `sync15::Error` and know what the underlying error was we'd be in a bit of trouble, but `anyhow::Error` probably makes that easier anyway.

Where this *does* fall apart is a possible future where the sync15 engines needed to call a sync15 implemented function as part of doing the sync, and that error needs to be inspected by the sync15 caller of the engine trait (phew!) - eg, if the design of the engine was such that an engine would explicitly request a http request be made. But the design of the engine avoids that - eg, the http requests are all initiated by the sync process and the engine is provided with the result.

Reply:
Aaah yeah that makes since - thanks @mhammond !!